p2p/src/peers/mod.rs

```diff
@@ -213,39 +214,60 @@ impl Peers {
     }
 
     /// Get a random socket address from the peers list
-    pub fn get_random(&self) -> Result<Option<SocketAddr>, failure::Error> {
-        let bucket = match (self.new_bucket.is_empty(), self.tried_bucket.is_empty()) {
-            (true, true) => return Ok(None),
-            (true, false) => &self.tried_bucket,
-            (false, true) => &self.new_bucket,
-            (false, false) => {
-                if thread_rng().gen() {
-                    &self.tried_bucket
-                } else {
-                    &self.new_bucket
-                }
-            }
-        };
+    /// This method provides the same probability to tried and new bucket peers
+    pub fn get_random_peers(&self, n: usize) -> Result<Vec<SocketAddr>, failure::Error> {
+        let mut rng = rand::thread_rng();
 
-        // Random index with range [0, len) of the peers vector
-        let index = thread_rng().gen_range(0, bucket.len());
+        let tried_len = self.tried_bucket.len();
+        let new_len = self.new_bucket.len();
 
-        Ok(bucket.values().nth(index).map(|v| v.address.to_owned()))
-    }
+        // Upper limit for this method is the sum of the two buckets length
+        let n_peers = cmp::min(tried_len + new_len, n);
 
-    /// Get a random socket address from the new peers list
-    pub fn get_new_random(&self) -> Option<(u16, SocketAddr)> {
-        if self.new_bucket.is_empty() {
-            return None;
+        // In case of 0 peers required, returns an empty vector
+        let mut v_peers: Vec<SocketAddr> = vec![];
+        if n_peers == 0 {
+            return Ok(v_peers);
         }
 
-        // Random index with range [0, len) of the peers vector
-        let index = thread_rng().gen_range(0, self.new_bucket.len());
+        // Calculate threshold between new and tried peers required
+        let threshold = rng.gen_range(0, n_peers);
```

Comment:
If we want 50% probability of each bucket, we need to roll a dice n times.

```suggestion
        // Run n experiments with probability of success 50%
        let threshold = (0..n_peers).fold(0, |acc, _| acc + rng.gen_range(0, 1));
```